doc/developer/design/20230330_recursion_limit.md

```diff
@@ -0,0 +1,187 @@
+
+We should add the limit setting to EXPLAIN, so that tests can verify that the limit settings are correctly picked up (and not messed up by optimizer transforms if we decide to have separate limits for each WMR clause).
+
+We'll need to test that the hard and soft limits do the correct thing when reaching the limit.
+
+We should keep in mind that some optimizer transforms perform loops when analyzing and transforming a `LetRec`, and in these cases they should respect the limits.
+
+We should test that the default hard limit is low enough that reaching it doesn't take forever. This should be done with multiple worker threads as well, and ideally also on staging with actual network communication, because communication between workers can significantly raise the per-iteration overhead (see the "Iteration Step Overhead" section and Fig. 8. [in this paper](https://www.researchgate.net/profile/Gabor-Gevay/publication/349768477_Efficient_Control_Flow_in_Dataflow_Systems_When_Ease-of-Use_Meets_High_Performance/links/6040fe4a4585154e8c77ceb8/Efficient-Control-Flow-in-Dataflow-Systems-When-Ease-of-Use-Meets-High-Performance.pdf).). In small clusters, 1000 iterations is almost certainly fine.
+
+When we later decide to have per-WMR limits rather than per-query, then we'll have to have several tests that involve multiple WMR clauses in various relative positions (independent, sequenced, nested).
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+A default limit for 1. has the drawback that not only buggy queries might hit it, but some genuine ones also, as explained [above](#default-values). However, I think a little annoyance for some advanced WMR users is worth it to avoid annoyance for a large number of beginner WMR users.
+
+Going with a per-WMR limit has the drawback that it is non-trivial to implement and would introduce a maintenance burden. Therefore, I propose to have just one limit for an entire query for now, and consider a per-WMR limit as [future work](#future-work).
+
+As mentioned above, I think we should postpone use case 3. until we have more evidence that it is actually needed, to avoid unjustified implementation effort.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+- It would be good to know how soon we'll have proper dataflow cancellation (see [#2392](https://github.com/MaterializeInc/materialize/issues/2392), plus [here](https://materializeinc.slack.com/archives/C02PPB50ZHS/p1680179894283039?thread_ts=1680177489.365979&cid=C02PPB50ZHS) and [here](https://materializeinc.slack.com/archives/C02PPB50ZHS/p1680199694602249)). However, I think we should already implement a basic solution for both use cases 1. and 2., and later revisit the default setting of the hard limit when we have proper dataflow cancellation.
+- How large is the iteration overhead on a real cluster with multiple worker threads? If small, maybe we should set the default hard limit to 10000 instead of 1000?
```

Comment:
2. and 3. are similar, but the main difference is that in the debugging use case (2.), you only want to temporarily change the value, i.e., during a debugging session, so including it in the SQL definition doesn't matter much. This means that a session variable might have been ok for it. (This point is not so important anymore, because we moved away from the session/system variables approach, because a system variable was deemed not suitable for use case 1., and I'd say if we have a SQL extension for 1., then it's easy to also have that for 2.)

Reply:
Another difference is that a broadcast join is kind of ok when just debugging, but it would be bad for 3.